addons/website_profile/controllers/main.py

```diff
@@ -0,0 +1,144 @@
+            'header': kwargs.get('header', dict()),
+            'searches': kwargs.get('searches', dict()),
+            'validation_email_sent': request.session.get('validation_email_sent', False),
+            'validation_email_done': request.session.get('validation_email_done', False),
+        }
+        values.update(kwargs)
+        return values
+
+    def _prepare_user_profile_parameters(self, **post):
+        return post
+
+    def _prepare_user_profile_values(self, user, **post):
+        return {
+            'uid': request.env.user.id,
+            'user': user,
+            'main_object': user,
+            'is_profile_page': True,
+            'edit_button_url_param': '',
+        }
+
+    @http.route(['/profile/user/<int:user_id>'], type='http', auth="public", website=True)
+    def view_user_profile(self, user_id, **post):
+        user = self._check_user_profile_access(user_id)
+        if not user:
+            return request.render("website_profile.private_profile", {}, status=404)
+        params = self._prepare_user_profile_parameters(**post)
+        values = self._prepare_user_profile_values(user, **params)
+        return request.render("website_profile.user_profile_main", values)
+
+    # Edit Profile
+    # ---------------------------------------------------
+    @http.route('/profile/edit', type='http', auth="user", website=True)
+    def view_user_profile_edition(self, **kwargs):
+        countries = request.env['res.country'].search([])
+        values = self._prepare_user_values(searches=kwargs)
+        values.update({
+            'email_required': kwargs.get('email_required'),
+            'countries': countries,
+            'notifications': self._get_badge_granted_messages(),
+            'url_param': kwargs.get('url_param'),
+        })
+        return request.render("website_profile.user_profile_edit_main", values)
+
+    def _profile_edition_preprocess_values(self, user, **kwargs):
+        values = {
+            'name': kwargs.get('name'),
+            'website': kwargs.get('website'),
+            'email': kwargs.get('email'),
+            'city': kwargs.get('city'),
+            'country_id': int(kwargs.get('country')) if kwargs.get('country') else False,
+            'website_description': kwargs.get('description'),
+        }
+
+        if 'clear_image' in kwargs:
+            values['image'] = False
+        elif kwargs.get('ufile'):
+            image = kwargs.get('ufile').read()
+            values['image'] = base64.b64encode(image)
+
+        if request.uid == user.id:  # the controller allows to edit only its own privacy settings; use partner management for other cases
+            values['website_published'] = kwargs.get('website_published') == 'True'
+        return values
+
+    def _save_edited_profile(self, **kwargs):
+        user = request.env.user
+        values = self._profile_edition_preprocess_values(user, **kwargs)
+        user.write(values)
```

Comment:
As we give `kwagrs` from user input as write parameter, I just wonder if the `SELF_WRITABLE_FIELDS` mecanisn is enough to secure this flow. Is it dangerous if someone add more values here (like livechat_username, voip config fields, ... that are in the self-writable list), or should we restrict the fields from kwargs ? @tde-banana-odoo You are the security expert of the team. Need your advice here.

Reply:
To be sure I would white list them, as discussed.